### PR TITLE
[FIXED JENKINS-38005] Properly log failure due to empty archive in Pipeline

### PR DESCRIPTION
**Pull request by abayer**
# Description

See [JENKINS-38005](https://issues.jenkins-ci.org/browse/JENKINS-38005).

Details: Since the job result is probably still null at the time that archiving
occurs, we weren't actually logging anything in Pipeline jobs in that
case. That was not ideal. So instead, let's log things, and also let's
update the error message to include a tip to use the allowEmptyArchive
option.

Note that there is no test for this added - I couldn't figure out a way to reproduce the Pipeline use case here, i.e., a still-null `result` when we get to `ArtifactArchiver`.

### Changelog entries

Proposed changelog entries:

* Entry 1: Properly log failure due to empty archive in Pipeline

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/resources/hudson/tasks/Messages.properties [hunk 1]
```diff
@@ -36,7 +36,7 @@ ArtifactArchiver.NoIncludes=\
 No artifacts are configured for archiving.\n\
 You probably forgot to set the file pattern, so please go back to the configuration and specify it.\n\
 If you really did mean to archive all the files in the workspace, please specify "**"
-ArtifactArchiver.NoMatchFound=No artifacts found that match the file pattern "{0}". Configuration error?
+ArtifactArchiver.NoMatchFound=No artifacts found that match the file pattern "{0}". Did you mean to use the "allowEmptyArchive" option?
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/tasks/Messages.properties]: This will be shown in Freestyle jobs, where the new message makes no sense.

**Review by daniel-beck — CHANGES_REQUESTED** [review]

**Comment by daniel-beck** [comment]
> I couldn't figure out a way to reproduce the Pipeline use case here, i.e., a still-null `result` when we get to `ArtifactArchiver`.

Does it not work with no build steps before this?

**Comment by abayer** [comment]
@daniel-beck Aaaaaaah - why did I not think of that? Checking - thanks.

File: core/src/main/resources/hudson/tasks/Messages.properties (hunk 1 above)
Comment (abayer) [inline core/src/main/resources/hudson/tasks/Messages.properties]: Hrm. Good point. Maybe use a different error message depending on whether we're being called in an `AbstractBuild` or not?

**Comment by abayer** [comment]
@daniel-beck No, wait, still won't work - `result` is getting set before we ever get to post-build actions.

File: core/src/main/resources/hudson/tasks/Messages.properties (hunk 1 above)
Comment (jglick) [inline core/src/main/resources/hudson/tasks/Messages.properties]: Why does this not make sense in freestyle builds? Even if the problem is not the consequence of an earlier failure, you might simply have wanted to allow zero artifacts in some cases.

**Review by jglick — APPROVED** [review]
Not easily tested without a Pipeline dep. Could write a test in a downstream plugin repository I suppose.

File: core/src/main/resources/hudson/tasks/Messages.properties (hunk 1 above)
Comment (daniel-beck) [inline core/src/main/resources/hudson/tasks/Messages.properties]: @jglick I mean the specific label used. It's not called `allowEmptyArchive` on the UI.

Comment (sanjaydhawangale) [inline core/src/main/resources/hudson/tasks/Messages.properties]: exactly #2602

Comment (oleg-nenashev) [inline core/src/main/resources/hudson/tasks/Messages.properties]: @sanjaydhawangale WDYM?

Comment (daniel-beck) [inline core/src/main/resources/hudson/tasks/Messages.properties]: @oleg-nenashev Just look at the profile. Random spam.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Looks good to me, it will also help other plugins like AnyBuildStep plugin. Hence 🐝 

But the merge conflict needs to be solved

**Review by oleg-nenashev — APPROVED** [review]
:bee:

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Change to localized message still doesn't make sense.

File: core/src/main/resources/hudson/tasks/Messages.properties (hunk 1 above)
Comment (abayer) [inline core/src/main/resources/hudson/tasks/Messages.properties]: @daniel-beck Do you have suggestions for alternatives?

**Comment by abayer** [comment]
@daniel-beck What would you recommend?

**Comment by daniel-beck** [comment]
@abayer Not changing it. The existing message is sufficient, and independent of *how* the step is used.

**Comment by abayer** [comment]
@daniel-beck Gotcha - will do.

**Review by jglick — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
:bee: and @reviewbybees done

**Review by daniel-beck — APPROVED** [review]
Should probably be squashed given how trivial this ended up being.

**Comment by oleg-nenashev** [comment]
Will merge it tomorrow if nobody votes again.
P.S: I am a professional squash-merger anyway

**Comment by jglick** [comment]
BTW in [JENKINS-31931](https://issues.jenkins-ci.org/browse/JENKINS-31931) I suggested this fix, plus some additional cleanup.

**Pull request merged**